### ISSUE TITLE
Intermittent exception thrown in bundle install (TypeError: class or module required for rescue clause)

### ISSUE THREAD

**Issue by petenorth** [oracle/truffleruby#2130]
truffleruby 20.3.0-dev-1fb68a0e, like ruby 2.6.6, GraalVM CE Native [x86_64-linux]

```
TypeError: class or module required for rescue clause
  /truffleruby-head/lib/truffle/truffle/cext.rb:1503:in `rb_rescue2'
  exception.c:90:in `rb_rescue2'
  zlib.c:2409:in `gzfile_read_raw'
  zlib.c:2694:in `gzfile_read_more'
  zlib.c:2825:in `gzfile_read_all'
  zlib.c:3828:in `rb_gzreader_read'
  /truffleruby-head/lib/truffle/truffle/cext_ruby.rb:39:in `read'
  /truffleruby-head/lib/mri/bundler/compact_index_client/updater.rb:58:in `block in update'
  /truffleruby-head/lib/mri/tmpdir.rb:93:in `mktmpdir'
  /truffleruby-head/lib/mri/bundler/compact_index_client/updater.rb:31:in `update'
  /truffleruby-head/lib/mri/bundler/compact_index_client.rb:82:in `update'
  /truffleruby-head/lib/mri/bundler/compact_index_client.rb:69:in `update_and_parse_checksums!'
  /truffleruby-head/lib/mri/bundler/fetcher/compact_index.rb:69:in `available?'
  /truffleruby-head/lib/mri/bundler/fetcher/compact_index.rb:16:in `call'
  /truffleruby-head/lib/mri/bundler/fetcher/compact_index.rb:16:in `available?'
  /truffleruby-head/lib/mri/bundler/fetcher.rb:158:in `use_api'
  /truffleruby-head/lib/mri/bundler/source/rubygems.rb:382:in `block in api_fetchers'
  /truffleruby-head/lib/mri/bundler/source/rubygems.rb:382:in `select'
  /truffleruby-head/lib/mri/bundler/source/rubygems.rb:382:in `api_fetchers'
  /truffleruby-head/lib/mri/bundler/source/rubygems.rb:387:in `block in remote_specs'
  /truffleruby-head/lib/mri/bundler/index.rb:11:in `build'
  /truffleruby-head/lib/mri/bundler/source/rubygems.rb:386:in `remote_specs'
  /truffleruby-head/lib/mri/bundler/source/rubygems.rb:89:in `specs'
  /truffleruby-head/lib/mri/bundler/definition.rb:273:in `block (2 levels) in index'
  /truffleruby-head/lib/mri/bundler/definition.rb:271:in `each'
  /truffleruby-head/lib/mri/bundler/definition.rb:271:in `block in index'
  /truffleruby-head/lib/mri/bundler/index.rb:11:in `build'
  /truffleruby-head/lib/mri/bundler/definition.rb:268:in `index'
  /truffleruby-head/lib/mri/bundler/definition.rb:258:in `resolve'
  /truffleruby-head/lib/mri/bundler/definition.rb:170:in `specs'
  /truffleruby-head/lib/mri/bundler/definition.rb:158:in `resolve_remotely!'
  /truffleruby-head/lib/mri/bundler/installer.rb:310:in `resolve_if_needed'
  /truffleruby-head/lib/mri/bundler/installer.rb:84:in `block in run'
  /truffleruby-head/lib/mri/bundler/process_lock.rb:12:in `block in lock'
  <internal:core> core/io.rb:688:in `open'
  /truffleruby-head/lib/mri/bundler/process_lock.rb:9:in `lock'
  /truffleruby-head/lib/mri/bundler/installer.rb:73:in `run'
  /truffleruby-head/lib/mri/bundler/installer.rb:25:in `install'
  /truffleruby-head/lib/mri/bundler/cli/install.rb:65:in `run'
  /truffleruby-head/lib/mri/bundler/cli.rb:235:in `block in install'
  /truffleruby-head/lib/mri/bundler/settings.rb:143:in `temporary'
  /truffleruby-head/lib/mri/bundler/cli.rb:234:in `install'
  /truffleruby-head/lib/mri/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
  /truffleruby-head/lib/mri/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
  /truffleruby-head/lib/mri/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
  /truffleruby-head/lib/mri/bundler/cli.rb:27:in `dispatch'
  /truffleruby-head/lib/mri/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
  /truffleruby-head/lib/mri/bundler/cli.rb:18:in `start'
  /truffleruby-head/lib/gems/gems/bundler-1.17.2/exe/bundle:30:in `block in <top (required)>'
  /truffleruby-head/lib/mri/bundler/friendly_errors.rb:124:in `with_friendly_errors'
  /truffleruby-head/lib/gems/gems/bundler-1.17.2/exe/bundle:22:in `<top (required)>'
  <internal:core> core/kernel.rb:401:in `load'
  <internal:core> core/kernel.rb:401:in `load'
  /truffleruby-head/bin/bundle:43:in `<main>'
```

Full github workflow

https://github.com/petenorth/schools-experience/runs/1299590958?check_suite_focus=true

**Comment by eregon** [comment]
I think it's the same as https://github.com/oracle/truffleruby/issues/2127#issuecomment-713741999
I'll try to fix it soon.

**Comment by eregon** [comment]
Fixed in 953cc2173723074a388a01fe12cf8ea39094ed0f, thanks for the report.
(the spec for it unfortunately checked for a TypeError going through, and we raised another TypeError due to the bug, which got unnoticed)
